provider/datagen/src/driver.rs

```diff
@@ -950,88 +931,111 @@ fn select_locales_for_key(
-
-    // TODO(#4629): Allow the exclusion of und
-    let include_und = true;
-
-    let result = supported_and_explicit
-        .filter(|locale_orig| {
-            let mut locale = locale_orig.clone();
-            locale.remove_aux();
-            if implicit.contains(&locale) {
-                return true;
+    // Add the explicit langids to the map
+    let mut include_full = false;
+    match locales_fallback {
+        LocalesWithOrWithoutFallback::WithFallback { families, .. } => {
+            if families.is_empty() {
+                // If no locales are selected but fallback is enabled, select the root locale
+                let value = locales_map.entry(LanguageIdentifier::UND).or_default();
+                value.is_selected = true;
             }
-            if explicit.contains(&locale) {
-                return true;
+            for family in families {
+                let Some(langid) = family.langid.as_ref() else {
+                    debug_assert_eq!(*family, LocaleFamily::full());
+                    include_full = true;
+                    continue;
+                };
+                let value = locales_map.entry(langid.clone()).or_default();
+                value.is_selected = true;
+                if *langid != LanguageIdentifier::UND {
+                    value.family = Some(family);
+                }
             }
-            if locale.is_langid_und() && include_und {
-                return true;
+            if include_full && families.len() == 1 {
+                // Special case: return now so we don't need the fallbacker
+                let selected_locales = locales_map
+                    .into_iter()
+                    .flat_map(|(_, value)| value.data_locales)
+                    .collect();
+                return Ok(selected_locales);
             }
-            if locale.language().is_empty()
-                && matches!(
-                    key.fallback_config().priority,
-                    icu_provider::FallbackPriority::Region
-                )
-            {
-                return true;
+        }
+        LocalesWithOrWithoutFallback::WithoutFallback { langids: locales } => {
+            for langid in locales {
+                let value = locales_map.entry(langid.clone()).or_default();
+                value.is_selected = true;
             }
-            // Special case: skeletons *require* the -u-ca keyword, so don't export locales that don't have it
-            // This would get caught later on, but it makes datagen faster and quieter to catch it here
-            if key.path().get() == "datetime/skeletons@1" && !locale.has_unicode_ext() {
-                return false;
+        }
+    }
+
+    // Fill in missing extensions and aux keys from parent locales,
+    // and calculate which langids are ancestors and descendants.
+    for current_langid in locales_map.keys().cloned().collect::<Vec<_>>() {
```

Comment:
There's a lot of cloning going on here because otherwise I would need multiple mutable borrows of the HashMap at the same time.